### ISSUE TITLE
New compilation warnings with gcc 8.3.0

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#17013]
Migrated from [rt.perl.org#134129](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=134129) (status was 'open')

Searchable as RT134129$

**Comment by p5pRT** [comment]
### From @dur-randir

#### Created by @dur-randir



I get the following warnings when building perl on Debian with gcc
8\.3\.0\. Wimplicit\-fallthrough is from at least 7\.4\.0\, others are new
ones&#8203;:

Wcast\-function\-type \(some of them\, all are about signal handlers\)&#8203;:

perl\.c&#8203;: In function ‘Perl\_sys\_init’&#8203;:
perl\.h&#8203;:2688&#8203;:51&#8203;: warning&#8203;: cast between incompatible function types from
‘\_\_sighandler\_t’ \{aka ‘void \(\*\)\(int\)’\} to ‘void \(\*\)\(int\,  siginfo\_t \*\,
void \*\)’ \{aka ‘void \(\*\)\(int\,  struct \<anonymous> \*\, void \*\)’\}
\[\-Wcast\-function\-type\]
 \#    define PERL\_FPU\_INIT       PL\_sigfpe\_saved = \(Sighandler\_t\)
signal\(SIGFPE\, SIG\_IGN\)
&nbsp;                                                   ^
unixish\.h&#8203;:150&#8203;:29&#8203;: note&#8203;: in expansion of macro ‘PERL\_FPU\_INIT’
&nbsp;  MALLOC\_CHECK\_TAINT2\(\*c\,\*v\) PERL\_FPU\_INIT; PERLIO\_INIT; MALLOC\_INIT
&nbsp;                             ^~~~~~~~~~~~~
perl\.c&#8203;:125&#8203;:5&#8203;: note&#8203;: in expansion of macro ‘PERL\_SYS\_INIT\_BODY’
&nbsp;     PERL\_SYS\_INIT\_BODY\(argc\, argv\);
&nbsp;     ^~~~~~~~~~~~~~~~~~

perl\.c&#8203;: In function ‘S\_parse\_body’&#8203;:
perl\.c&#8203;:2399&#8203;:22&#8203;: warning&#8203;: cast between incompatible function types from
‘void \(\*\)\(int\)’ to ‘void \(\*\)\(int\,  siginfo\_t \*\, void \*\)’ \{aka ‘void
\(\*\)\(int\,  struct \<anonymous> \*\, void \*\)’\} \[\-Wcast\-function\-type\]
&nbsp;      if \(sigstate == \(Sighandler\_t\) SIG\_IGN\) \{
&nbsp;                      ^
In file included from perl\.h&#8203;:5359\,
&nbsp;                 from perl\.c&#8203;:37&#8203;:
perl\.c&#8203;:2402&#8203;:26&#8203;: warning&#8203;: cast between incompatible function types from
‘void \(\*\)\(int\)’ to ‘void \(\*\)\(int\,  siginfo\_t \*\, void \*\)’ \{aka ‘void
\(\*\)\(int\,  struct \<anonymous> \*\, void \*\)’\} \[\-Wcast\-function\-type\]
&nbsp;   \(void\)rsignal\(SIGCHLD\, \(Sighandler\_t\)SIG\_DFL\);
&nbsp;                          ^
embed\.h&#8203;:643&#8203;:44&#8203;: note&#8203;: in definition of macro ‘rsignal’
 \#define rsignal\(a\,b\)  Perl\_rsignal\(aTHX\_ a\,b\)

util\.c&#8203;: In function ‘Perl\_rsignal’&#8203;:
util\.c&#8203;:2698&#8203;:22&#8203;: warning&#8203;: cast between incompatible function types from
‘Sighandler\_t’ \{aka ‘void \(\*\)\(int\,  struct \<anonymous> \*\, void \*\)’\} to
‘void \(\*\)\(int\)’ \[\-Wcast\-function\-type\]
&nbsp;     act\.sa\_handler = \(void\(\*\)\(int\)\)handler;
&nbsp;                      ^

pp\_sys\.c&#8203;: In function ‘Perl\_pp\_system’&#8203;:
pp\_sys\.c&#8203;:4421&#8203;:28&#8203;: warning&#8203;: cast between incompatible function types
from ‘void \(\*\)\(int\)’ to ‘void \(\*\)\(int\,  siginfo\_t \*\, void \*\)’ \{aka
‘void \(\*\)\(int\,  struct \<anonymous> \*\, void \*\)’\} \[\-Wcast\-function\-type\]
&nbsp;      rsignal\_save\(SIGINT\,  \(Sighandler\_t\) SIG\_IGN\, &ihand\);
&nbsp;                            ^
embed\.h&#8203;:1560&#8203;:55&#8203;: note&#8203;: in definition of macro ‘rsignal\_save’
 \#define rsignal\_save\(a\,b\,c\) Perl\_rsignal\_save\(aTHX\_ a\,b\,c\)
&nbsp;                                                       ^
Wimplicit\-fallthrough&#8203;:

decompress\.c&#8203;: In function ‘BZ2\_decompress’&#8203;:
decompress\.c&#8203;:198&#8203;:10&#8203;: warning&#8203;: this statement may fall through
\[\-Wimplicit\-fallthrough=\]
&nbsp;       if \(uc \!= BZ\_HDR\_B\) RETURN\(BZ\_DATA\_ERROR\_MAGIC\);
&nbsp;          ^
decompress\.c&#8203;:44&#8203;:4&#8203;: note&#8203;: here
&nbsp;    case lll&#8203;: s\->state = lll;                      \\
&nbsp;    ^~~~
decompress\.c&#8203;:68&#8203;:4&#8203;: note&#8203;: in expansion of macro ‘GET\_BITS’
&nbsp;    GET\_BITS\(lll\,uuu\,8\)
&nbsp;    ^~~~~~~~
decompress\.c&#8203;:200&#8203;:7&#8203;: note&#8203;: in expansion of macro ‘GET\_UCHAR’
&nbsp;       GET\_UCHAR\(BZ\_X\_MAGIC\_2\, uc\);
&nbsp;       ^~~~~~~~~

Wmultistatement\-macros&#8203;:

deflate\.c&#8203;: In function ‘deflateParams’&#8203;:
deflate\.c&#8203;:602&#8203;:28&#8203;: warning&#8203;: macro expands to multiple statements
\[\-Wmultistatement\-macros\]
&nbsp;                 CLEAR\_HASH\(s\);
&nbsp;                            ^
deflate\.c&#8203;:193&#8203;:5&#8203;: note&#8203;: in definition of macro ‘CLEAR\_HASH’
&nbsp;     s\->head\[s\->hash\_size\-1\] = NIL; \\
&nbsp;     ^
deflate\.c&#8203;:601&#8203;:13&#8203;: note&#8203;: some parts of macro expansion are not guarded
by this ‘else’ clause
&nbsp;             else
&nbsp;             ^~~~

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.31.0:

Configured by dur-randir at Wed Feb 27 14:51:01 MSK 2019.

Summary of my perl5 (revision 5 version 31 subversion 0) configuration:
  Commit id: 58f4626762668e1c1948832073998af84b2c85d0
  Platform:
    osname=linux
    osvers=4.19.0-5-amd64
    archname=x86_64-linux
    uname='linux dorothy 4.19.0-5-amd64 #1 smp debian 4.19.37-3
(2019-05-15) x86_64 gnulinux '
    config_args='-de -Dusedevel -DDEBUGGING -Dcc=gcc-8'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='gcc-8'
    ccflags ='-fwrapv -DDEBUGGING -fno-strict-aliasing -pipe
-fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE
-D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2 -g'
    cppflags='-fwrapv -DDEBUGGING -fno-strict-aliasing -pipe
-fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='8.3.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='gcc-8'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib
/usr/lib/gcc/x86_64-linux-gnu/8/include-fixed
/usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu
/lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.28.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.28'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl):
  Compile-time options:
    DEBUGGING
    HAS_TIMES
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
  Built under linux
  Compiled at May 23 2019 11:40:52
  %ENV:
    PERLBREW_BASHRC_VERSION="0.78"
    PERLBREW_HOME="/home/afl/.perlbrew"
    PERLBREW_MANPATH="/home/afl/perlbrew/perls/perl-5.20.2/man"
    PERLBREW_PATH="/home/afl/perlbrew/bin:/home/afl/perlbrew/perls/perl-5.20.2/bin"
    PERLBREW_PERL="perl-5.20.2"
    PERLBREW_ROOT="/home/afl/perlbrew"
    PERLBREW_VERSION="0.78"
  @INC:
    lib
    /usr/local/lib/perl5/site_perl/5.31.0/x86_64-linux
    /usr/local/lib/perl5/site_perl/5.31.0
    /usr/local/lib/perl5/5.31.0/x86_64-linux
    /usr/local/lib/perl5/5.31.0

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Thu\, 23 May 2019 09&#8203;:13&#8203;:12 GMT\, randir wrote&#8203;:
> This is a bug report for perl from sergey\.aleynikov@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.41 running under perl 5\.31\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> I get the following warnings when building perl on Debian with gcc
> 8\.3\.0\. Wimplicit\-fallthrough is from at least 7\.4\.0\, others are new
> ones&#8203;:
> 
> Wcast\-function\-type \(some of them\, all are about signal handlers\)&#8203;:
> 
> perl\.c&#8203;: In function ‘Perl\_sys\_init’&#8203;:
> perl\.h&#8203;:2688&#8203;:51&#8203;: warning&#8203;: cast between incompatible function types from
> ‘\_\_sighandler\_t’ \{aka ‘void \(\*\)\(int\)’\} to ‘void \(\*\)\(int\,  siginfo\_t \*\,
> void \*\)’ \{aka ‘void \(\*\)\(int\,  struct \<anonymous> \*\, void \*\)’\}
> \[\-Wcast\-function\-type\]
>  \#    define PERL\_FPU\_INIT       PL\_sigfpe\_saved = \(Sighandler\_t\)
> signal\(SIGFPE\, SIG\_IGN\)
>                                                    ^
> unixish\.h&#8203;:150&#8203;:29&#8203;: note&#8203;: in expansion of macro ‘PERL\_FPU\_INIT’
>   MALLOC\_CHECK\_TAINT2\(\*c\,\*v\) PERL\_FPU\_INIT; PERLIO\_INIT; MALLOC\_INIT
>                              ^~~~~~~~~~~~~
> perl\.c&#8203;:125&#8203;:5&#8203;: note&#8203;: in expansion of macro ‘PERL\_SYS\_INIT\_BODY’
>      PERL\_SYS\_INIT\_BODY\(argc\, argv\);
>      ^~~~~~~~~~~~~~~~~~
> 
> perl\.c&#8203;: In function ‘S\_parse\_body’&#8203;:
> perl\.c&#8203;:2399&#8203;:22&#8203;: warning&#8203;: cast between incompatible function types from
> ‘void \(\*\)\(int\)’ to ‘void \(\*\)\(int\,  siginfo\_t \*\, void \*\)’ \{aka ‘void
> \(\*\)\(int\,  struct \<anonymous> \*\, void \*\)’\} \[\-Wcast\-function\-type\]
>       if \(sigstate == \(Sighandler\_t\) SIG\_IGN\) \{
>                       ^
> In file included from perl\.h&#8203;:5359\,
>                  from perl\.c&#8203;:37&#8203;:
> perl\.c&#8203;:2402&#8203;:26&#8203;: warning&#8203;: cast between incompatible function types from
> ‘void \(\*\)\(int\)’ to ‘void \(\*\)\(int\,  siginfo\_t \*\, void \*\)’ \{aka ‘void
> \(\*\)\(int\,  struct \<anonymous> \*\, void \*\)’\} \[\-Wcast\-function\-type\]
>    \(void\)rsignal\(SIGCHLD\, \(Sighandler\_t\)SIG\_DFL\);
>                           ^
> embed\.h&#8203;:643&#8203;:44&#8203;: note&#8203;: in definition of macro ‘rsignal’
>  \#define rsignal\(a\,b\)  Perl\_rsignal\(aTHX\_ a\,b\)
> 
> util\.c&#8203;: In function ‘Perl\_rsignal’&#8203;:
> util\.c&#8203;:2698&#8203;:22&#8203;: warning&#8203;: cast between incompatible function types from
> ‘Sighandler\_t’ \{aka ‘void \(\*\)\(int\,  struct \<anonymous> \*\, void \*\)’\} to
> ‘void \(\*\)\(int\)’ \[\-Wcast\-function\-type\]
>      act\.sa\_handler = \(void\(\*\)\(int\)\)handler;
>                       ^
> 
> pp\_sys\.c&#8203;: In function ‘Perl\_pp\_system’&#8203;:
> pp\_sys\.c&#8203;:4421&#8203;:28&#8203;: warning&#8203;: cast between incompatible function types
> from ‘void \(\*\)\(int\)’ to ‘void \(\*\)\(int\,  siginfo\_t \*\, void \*\)’ \{aka
> ‘void \(\*\)\(int\,  struct \<anonymous> \*\, void \*\)’\} \[\-Wcast\-function\-type\]
>       rsignal\_save\(SIGINT\,  \(Sighandler\_t\) SIG\_IGN\, &ihand\);
>                             ^
> embed\.h&#8203;:1560&#8203;:55&#8203;: note&#8203;: in definition of macro ‘rsignal\_save’
>  \#define rsignal\_save\(a\,b\,c\) Perl\_rsignal\_save\(aTHX\_ a\,b\,c\)
>                                                        ^
> Wimplicit\-fallthrough&#8203;:
> 
> decompress\.c&#8203;: In function ‘BZ2\_decompress’&#8203;:
> decompress\.c&#8203;:198&#8203;:10&#8203;: warning&#8203;: this statement may fall through
> \[\-Wimplicit\-fallthrough=\]
>        if \(uc \!= BZ\_HDR\_B\) RETURN\(BZ\_DATA\_ERROR\_MAGIC\);
>           ^
> decompress\.c&#8203;:44&#8203;:4&#8203;: note&#8203;: here
>     case lll&#8203;: s\->state = lll;                      \\
>     ^~~~
> decompress\.c&#8203;:68&#8203;:4&#8203;: note&#8203;: in expansion of macro ‘GET\_BITS’
>     GET\_BITS\(lll\,uuu\,8\)
>     ^~~~~~~~
> decompress\.c&#8203;:200&#8203;:7&#8203;: note&#8203;: in expansion of macro ‘GET\_UCHAR’
>        GET\_UCHAR\(BZ\_X\_MAGIC\_2\, uc\);
>        ^~~~~~~~~
> 
> Wmultistatement\-macros&#8203;:
> 
> deflate\.c&#8203;: In function ‘deflateParams’&#8203;:
> deflate\.c&#8203;:602&#8203;:28&#8203;: warning&#8203;: macro expands to multiple statements
> \[\-Wmultistatement\-macros\]
>                  CLEAR\_HASH\(s\);
>                             ^
> deflate\.c&#8203;:193&#8203;:5&#8203;: note&#8203;: in definition of macro ‘CLEAR\_HASH’
>      s\->head\[s\->hash\_size\-1\] = NIL; \\
>      ^
> deflate\.c&#8203;:601&#8203;:13&#8203;: note&#8203;: some parts of macro expansion are not guarded
> by this ‘else’ clause
>              else
>              ^~~~
> 

Thank you for this analysis\.  Some of these warnings are from modules that are maintained upstream on CPAN\.  In the case of\, say\, decompress\.c\, that's part of IO\-Compress\, whose maintainer has previously indicated that he modifies that code if and when the underlying C libraries get updated\.  So we \(p5p\) can't directly control that\.

However\, if you are able to submit patches for the warnings being emitted by those files maintained by p5p \(e\.g\.\, perl\.c\, util\.c\)\, we would be happy to consider them\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by jkeenan** [comment]
> ### From @jkeenan
> 
> On Thu, 23 May 2019 09​:13​:12 GMT, randir wrote​:
> 
[snip]
> > I get the following warnings when building perl on Debian with gcc
> > 8.3.0. Wimplicit-fallthrough is from at least 7.4.0, others are new
> > ones​:
> > Wcast-function-type (some of them, all are about signal handlers)​:

[snip]
> 
> Thank you for this analysis. Some of these warnings are from modules that are maintained upstream on CPAN. In the case of, say, decompress.c, that's part of IO-Compress, whose maintainer has previously indicated that he modifies that code if and when the underlying C libraries get updated. So we (p5p) can't directly control that.
> 
> However, if you are able to submit patches for the warnings being emitted by those files maintained by p5p (e.g., perl.c, util.c), we would be happy to consider them.
> 
> Thank you very much.
> 
> --
> James E Keenan (jkeenan@​cpan.org)



For anyone who wants to take a go at clearing up some of these warnings, please see the attachment which lists all the Wcast-function-type warnings, all of which are in files under p5p maintenance.

Thank you very much.
Jim Keenan
[make.0463f3a1.linux.gcc-8.Wcast-function-type.warnings.txt](https://github.com/Perl/perl5/files/3853678/make.0463f3a1.linux.gcc-8.Wcast-function-type.warnings.txt)

**Comment by xenu** [comment]
@iabyn is working on this, see https://www.nntp.perl.org/group/perl.perl5.porters/2019/11/msg256570.html

**Comment by jkeenan** [comment]
> @iabyn is working on this, see https://www.nntp.perl.org/group/perl.perl5.porters/2019/11/msg256570.html

Thanks for letting me know this.  Yes, when that is merged, it will eliminate that particular group of warnings.


```
$ report-build-warnings make.0463f3a1.linux.gcc-8.output.txt.gz 
File:  make.0463f3a1.linux.gcc-8.output.txt.gz

  Wcast-function-type                       23
  Wimplicit-fallthrough=                    34
  Wmultistatement-macros                     1
  Wunused-but-set-variable                   1
  Wunused-parameter                          2
  Wunused-variable                           4

$ report-build-warnings make.sighandler2.linux.gcc-8.output.txt.gz 
File:  make.sighandler2.linux.gcc-8.output.txt.gz

  Wimplicit-fallthrough=                    34
  Wmultistatement-macros                     1
  Wunused-but-set-variable                   1
  Wunused-parameter                          2
  Wunused-variable                           4
```

**Comment by pmqs** [comment]
> ### From @jkeenan
> On Thu, 23 May 2019 09​:13​:12 GMT, randir wrote​:
> 
> > This is a bug report for perl from sergey.aleynikov@​gmail.com,
> > generated with the help of perlbug 1.41 running under perl 5.31.0.
> > -----------------------------------------------------------------
> > [Please describe your issue here]
> > I get the following warnings when building perl on Debian with gcc
> > 8.3.0. Wimplicit-fallthrough is from at least 7.4.0, others are new
> > ones​:
...
> 
> Thank you for this analysis. Some of these warnings are from modules that are maintained upstream on CPAN. In the case of, say, decompress.c, that's part of IO-Compress, whose maintainer has previously indicated that he modifies that code if and when the underlying C libraries get updated. So we (p5p) can't directly control that.

That had been my opinion a while back, but the upstream libraries I depend on, namely zlib and bzip2, have been very quiet (zlib will soon be three years since the last release, bzip2 was effectively dead until recently). I already include patches for Bzip2, so a few more aren't a problem.

> However, if you are able to submit patches for the warnings being emitted by those files maintained by p5p (e.g., perl.c, util.c), we would be happy to consider them.

And patches for bzip2 will also be considered. 

_**Just noticed that my email reply to this issue on P5P didn't make it into the github comments. Above is a cut-n-paste from P5P. Is this a bug/feature/limitation?**_

**Comment by Grinnz** [comment]
Email replies to perlbug-followup no longer go anywhere useful (I believe the mailbox is manually monitored).

**Comment by pmqs** [comment]
> Email replies to perlbug-followup no longer go anywhere useful (I believe the mailbox is manually monitored).

That's a shame

**Comment by pmqs** [comment]
I've updated `Compress-Raw-Zlib` and `Compress-Raw-Bzip2` to address all the warnings from this ticket. Fix is in version 2.091 of both modules.

**Comment by jkeenan** [comment]
@pmqs, thanks for your attention to this.  Your work has led to a dramatic reduction in the number of build-time warnings we're getting from gcc-8.
```
$ report-build-warnings make.d2e1c851.linux.gcc8.output.txt
File:  make.d2e1c851.linux.gcc8.output.txt

  Wmaybe-uninitialized                       2
  Wunused-variable                           4
```
(And none of the remaining warnings are in your scope.)

Thank you very much.
Jim Keenan
